### PR TITLE
Aiodns support

### LINKED ISSUE #728: Add optional support for aiodns to TCPConnector for async DNS resolving
See http://stackoverflow.com/questions/34700311/how-to-handle-dns-timeouts-with-aiohttp for reasoning

### PR DESCRIPTION
**Pull request by gwillem**
Fixes #728. Todo:

- [x] Mock DNS tests
- [x] Add docs

### CODE REVIEW

File: aiohttp/resolver.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import asyncio
+from .abc import AbstractResolver
+
+try:
+    import aiodns
+except ImportError:
+    aiodns = None
+
+
+class ExecutorResolver(AbstractResolver):
```
Comment (fafhrd91) [inline aiohttp/resolver.py]: Strictly speaking thread pool is just internal details of asyncio implementation, and may be change in future. I'd call it 'DefaultResolver' and document it as default asyncio based resolver

Comment (gwillem) [inline aiohttp/resolver.py]: Thanks!

File: tests/test_resolver.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import asyncio
```
Comment (asvetlov) [inline tests/test_resolver.py:2]: Please rewrite tests to use py.test style (like `test_client_connection.py` for example).

File: aiohttp/connector.py [hunk 3]
```diff
@@ -443,6 +447,13 @@ def __init__(self, *, verify_ssl=True, fingerprint=None,
         else:
             _use_dns_cache = False
 
+        self._resolver = resolver or DefaultResolver(loop=self._loop)
```
Comment (asvetlov) [inline aiohttp/connector.py:454]: Looks like you've missed `self._resolver.close()` call

**Comment by lphuberdeau** [comment]
Any way I can help this feature move forward?

**Comment by gwillem** [comment]
@lphuberdeau I was planning to continue on this PR in a week or 2-3 but I am absolutely swamped right now. You're most welcome to contribute if you have an earlier opportunity. Let me know.

**Comment by lphuberdeau** [comment]
@gwillem I've branched off your work. I'll notify you when the pull request is ready to be updated.

**Comment by lphuberdeau** [comment]
https://github.com/lphuberdeau/aiohttp/tree/aiodns-support
- I've converted the tests to py.test
- Mocked the DNS requests
- Added documentation

I'm a bit unclear about what needs to be tested in the TCPConnector without mocking absolutely everything.

About the missing .close(), it could be added, but given the object is instantiated outside of the connector, I'm not certain it is the connector's responsibility.

**Comment by gwillem** [comment]
Apparently pycares does not build on Win arch. Is this a showstopper, given that Aiodns is optional?

@lphuberdeau great work btw!

**Comment by lphuberdeau** [comment]
The tests will run fine without it as the skips are in place. Everything in pycares seem to indicate it should build on windows. To resolve the CI issue, I see two options:
- Separate the requirements file for travis and appveyor
- Move the aiodns installation to the travis file only

**Comment by gwillem** [comment]
FYI I have since migrated back to Tornado's AsyncHTTPClient for the following reasons:
- Uses c-ares2 async DNS resolving out of the box (I've tried adding aiodns in this PR)
- Does not suffer from cookie stacking/disclosure (#792)
- Uses common HTTP headers capitalization (#813)
- Is more fault tolerant with non-rfc compliant servers so is better suited for mass-crawling
- It is faster (in my setup) and leaks less memory.

IMHO aiohttp is cleaner, especially with native support for asyncio. And I would love to use aiohttp in the future. But right now I have deadlines :) so I need to be pragmatic.

**Comment by lphuberdeau** [comment]
@gwillem If you no longer want to track this issue, I can probably submit it back under a different PR and handle the back and forth.

**Comment by gwillem** [comment]
@lphuberdeau ok, please do! And again, thanks for joining in.

**Comment by lphuberdeau** [comment]
Feel free to close this one.

**Pull request closed without merging**